File: jscomp/test/ffi_arity_test.ml
external map : 'a array -> (('a -> 'b)[@bs]) -> 'b array = "map" [@@bs.send]

external mapi : 'a array -> (('a -> int -> 'b)[@bs]) -> 'b array = "map"
  [@@bs.send]

external parseInt : string -> int = "parseInt" [@@bs.val]
external parseInt_radix : string -> int -> int = "parseInt" [@@bs.val]

let f v = if v mod 2 = 0 then fun v -> v * v else fun v -> v + v
let v = mapi [|1; 2; 3|] (Js.Internal.fn_mk2 f)
let vv = mapi [|1; 2; 3|] (Js.Internal.fn_mk2 ( + ))
let hh = map [|"1"; "2"; "3"|] (Js.Internal.fn_mk1 parseInt)
let u = Js.Internal.fn_mk0 (fun _ -> 3)
let vvv = ref 0

let fff () =
  (* No inline *)
  Js.log "x" ; Js.log "x" ; incr vvv

let g = fun [@bs] () -> fff ()

(* will be compiled into var g = function () { fff (0)} not {[ var g = fff ]} *)
let abc x y z =
  Js.log "xx" ;
  Js.log "yy" ;
  x + y + z

let abc_u = fun [@bs] x y z -> abc x y z

(* cool, it will be compiled into {[ var absc_u = abc ]} *)
let () = (g () [@bs])

;;
Mt.from_pair_suites __MODULE__
  Mt.
    [ (__LOC__, fun _ -> Eq (v, [|0; 1; 4|]))
    ; (__LOC__, fun _ -> Eq (vv, [|1; 3; 5|]))
    ; (__LOC__, fun _ -> Eq (hh, [|1; 2; 3|]))
    ; ( __LOC__
      , fun _ ->
          Eq
            ( map
                (map [|1; 2; 3|] (Js.Internal.fn_mk1 (fun x y -> x + y)))
                (Js.Internal.fn_mk1 @@ fun y -> y 0 * y 1)
            , [|2; 6; 12|] ) )
    ; ( __LOC__
      , fun _ ->
          Eq
            ( mapi [|1; 2; 3|]
                (Js.Internal.fn_mk2 (fun x ->
                     let y = x * x in
                     fun i -> y + i))
            , [|1; 5; 11|] ) ) ]

(* FIXME: *)
let bar fn = fn ()

(* let hh0001 = fun%raw a b -> {| a + b|} let hh0002 = fun%raw () -> {|
   console.log ("forgiving arity")|} *)

;;
bar [%raw fun () -> {|console.log("forgiving arity")|}]

File: jscomp/test/ffi_arity_test.js
'use strict';

var Mt = require("./mt.js");
var Curry = require("../../lib/js/curry.js");

function f(v) {
  if (v % 2 === 0) {
    return function (v) {
      return Math.imul(v, v);
    };
  } else {
    return function (v) {
      return v + v | 0;
    };
  }
}

var v = [
    1,
    2,
    3
  ].map(function (a, b) {
      return f(a)(b);
    });

var vv = [
    1,
    2,
    3
  ].map(function (a, b) {
      return a + b | 0;
    });

var hh = [
    "1",
    "2",
    "3"
  ].map(function (x) {
      return parseInt(x);
    });

function u() {
  return 3;
}

var vvv = {
  contents: 0
};

function fff(param) {
  console.log("x");
  console.log("x");
  vvv.contents = vvv.contents + 1 | 0;
  
}

function g() {
  return fff(undefined);
}

function abc(x, y, z) {
  console.log("xx");
  console.log("yy");
  return (x + y | 0) + z | 0;
}

var abc_u = abc;

g();

Mt.from_pair_suites("Ffi_arity_test", {
      hd: [
        "File \"ffi_arity_test.ml\", line 45, characters 4-11",
        (function (param) {
            return {
                    TAG: /* Eq */0,
                    _0: v,
                    _1: [
                      0,
                      1,
                      4
                    ]
                  };
          })
      ],
      tl: {
        hd: [
          "File \"ffi_arity_test.ml\", line 46, characters 4-11",
          (function (param) {
              return {
                      TAG: /* Eq */0,
                      _0: vv,
                      _1: [
                        1,
                        3,
                        5
                      ]
                    };
            })
        ],
        tl: {
          hd: [
            "File \"ffi_arity_test.ml\", line 47, characters 4-11",
            (function (param) {
                return {
                        TAG: /* Eq */0,
                        _0: hh,
                        _1: [
                          1,
                          2,
                          3
                        ]
                      };
              })
          ],
          tl: /* [] */0
        }
      }
    });

function bar(fn) {
  return Curry._1(fn, undefined);
}

(Curry._1((function(){console.log("forgiving arity")}), undefined));

exports.f = f;
exports.v = v;
exports.vv = vv;
exports.hh = hh;
exports.u = u;
exports.vvv = vvv;
exports.fff = fff;
exports.g = g;
exports.abc = abc;
exports.abc_u = abc_u;
exports.bar = bar;
/* v Not a pure module */
